### ISSUE TITLE
Post/Pre deployment scripts support

### ISSUE THREAD

**Issue by molszews** [jmezach/MSBuild.Sdk.SqlProj#9]
Seems like these are not supported.

I've played a little bit with appending such scripts, seems like simple appending of the .sql to resulting dacpac (zip) should work (https://the.agilesql.club/2019/03/how-can-we-merge-multiple-dacpacs-into-one/), however, I'm not sure how to tackle sqlcmd syntax support.
Running `sqlcmd -i postdeployment_entryfile.sql -e -r >postdeploy.sql` generates correct file, but bundling sqlcmd within the target project may be difficult?

**Comment by ErikEJ** [comment]
Any thoughts about the design of this?

``` xml
  <ItemGroup>
    <PostDeploy Include="Postdeployment\Script.PostDeployment.sql" />
  </ItemGroup>

```
If this is passed to the command line tool, maybe we can just expect sqlcmd to be present, and run it to create the consolidated script(s), and add it like this: https://github.com/GoEddie/DacpacMerge/blob/master/src/MergeEm/Program.cs#L118

**Comment by molszews** [comment]
well, using sort of naming convention won't be enough? there can be only one post/pre deploy script, we can simply assume it should be available under `Post-deployment\postdeploy.sql`

**Comment by ErikEJ** [comment]
You can link multiple script with r: sqlcmd syntax - and we do.

**Comment by molszews** [comment]
yeah, but you specify only the entry point inside the .csproj, after that you can link every other file?
btw, sqlcmd could be embedded in a way this package embeds chromedriver binary https://github.com/jsakamoto/nupkg-selenium-webdriver-chromedriver/#where-is-chromedriverexe-saved-to

**Comment by ErikEJ** [comment]
But for v1, we could of course just go by convention, and expect the script to be merged.

**Comment by ErikEJ** [comment]
Not sure sqlcmd is simply a single .exe..

**Comment by jmezach** [comment]
I wonder how SSDT treats those scripts. I'm guessing it may just be added to the model as is and the passed to sqlcmd at deployment time. Does anyone have an example `.dacpac` that is build by SSDT which includes a pre- and/or postdeployment script that they are willing to share so I can have a look?

**Comment by molszews** [comment]
Resulting .dacpac archive contains a single .sql file which looks like the result of sqlcmd command I've posted before. It looks like there is nothing in the model related to that postdeploy.sql, it is simply appended to the zip archive. I will try to provide some samples

**Comment by ErikEJ** [comment]
If you email me, I can send you one privately.

**Comment by molszews** [comment]
@jmezach https://we.tl/t-ts0yOX4KA0

**Comment by ErikEJ** [comment]
I suggest that for initial implementation:

1: We assume the script is consolidated (the user can do this as a pipeline/build step)

2: We assume the folders follow the naming convention already specified here.

3: We assume the .sqlproj syntax will point to the consolidated script:
``` xml
<ItemGroup>
    <PostDeploy Include="Post-Deployment\Script.PostDeployment.sql" />
    <PreDeploy Include="Pre-Deployment\Script.PreDeployment.sql" />
  </ItemGroup>
```
Happy to update the command line tool to support this, if we decide to go ahead!

**Comment by ErikEJ** [comment]
I will implement like this: https://github.com/GoEddie/DacpacMerge/blob/f37ec0edbc1f77860bd745cd28d9c7d7030491e7/src/MergeEm/Program.cs#L112

**Comment by ErikEJ** [comment]
@jmezach Can I start work on this? Or are you already full speed ahead?

**Comment by jmezach** [comment]
Yeah, go for it :)

**Comment by jmezach** [comment]
@ErikEJ @molszews A new [beta package](https://www.nuget.org/packages/MSBuild.Sdk.SqlProj/1.1.0-beta.9) is now up on NuGet for you to try out. This should have support for pre/post deployment scripts, if everything is in the same file. I'm still working on the merge logic.

**Comment by ErikEJ** [comment]
@jmezach Fantastic! Will give it a try asap!

**Comment by ErikEJ** [comment]
![image](https://user-images.githubusercontent.com/4169187/81208068-d357c100-8fce-11ea-837d-73b29b653d92.png)

I am now able to build a "buddy" .dacpac to my .sqlproj project, and any differences in package content are not of importance. The postdeploy files even match!

My project file (not in the same folder as the .sqlproj) with linked items (so the .csproj is basically residing in an empty folder):

``` xml
<Project Sdk="MSBuild.Sdk.SqlProj/1.1.0-beta.11">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <SqlServerVersion>SqlAzure</SqlServerVersion>
    <AnsiNulls>True</AnsiNulls>
    <QuotedIdentifier>True</QuotedIdentifier>
    <CompatibilityMode>150</CompatibilityMode>   
  </PropertyGroup>

  <ItemGroup>
    <SqlCmdVariable Include="DbReaderPassword">
      <DefaultValue>
      </DefaultValue>
      <Value>$(SqlCmdVar__2)</Value>
    </SqlCmdVariable>
    <SqlCmdVariable Include="DbUserPassword">
      <DefaultValue>
      </DefaultValue>
      <Value>$(SqlCmdVar__1)</Value>
    </SqlCmdVariable>
  </ItemGroup>

  <ItemGroup>
    <Content Include="..\..\Database\dbo\**\*.sql" />
  </ItemGroup>
 
  <ItemGroup>
    <PostDeploy Include="..\..\Database\Post-Deployment\Script.PostDeployment.sql" />
  </ItemGroup>

</Project>

```

**Comment by jmezach** [comment]
Very nice. Glad to see it's coming together. Did you try deploying the resulting .dacpac yet? Curious to see if it actually runs at deployment time. @molszews Have you had time to give it a try yet?

I made some progress on the merging of scripts today. Interestingly it works fine on my dev machine, but fails on CI. For some reason it can't find the included file. Hope to do some more work on that soon.

**Comment by ErikEJ** [comment]
>>  Did you try deploying the resulting .dacpac yet? 

No, I need PR review etc. before I can deploy to DEV - will do so tomorrow

**Comment by ErikEJ** [comment]
I can now confirm, that deploying my package with both postdeploy and sqlcmd variables using Azure Pipelines to our DEV environment  works as expected. I believe #6 can be closed now.

**Comment by jmezach** [comment]
Great, thanks for reporting back. I'll close #6 when I've released a stable version. Do the pre/post-deployment scripts work as well?

**Comment by ErikEJ** [comment]
>> Do the pre/post-deployment scripts work as well?

Yes, it does! (Only using post)

**Comment by jmezach** [comment]
As I mentioned yesterday I did make some progress on the merging of SQLCMD files. I'm trying to rely on existing code from the Microsoft assemblies to do it, so that we don't have to reinvent the wheel, even though those API's are internal. Unfortunately it looks like the existing code isn't very cross-platform friendly. When you reference a script using something like `:r ./Script1.sql` it turns that into an absolute path, but also uppercases the entire path. Then it checks to see if that file exists, which obviously doesn't work on platforms that have a case sensitive filesystem.

If I run CI on Windows it will work just fine, but I would like for things to stay truly cross-platform. Would it be ok for you guys if I ship what we have right now as 1.1.0 and create a separate issue for the merging of files that may or may not come in a future version?

**Comment by ErikEJ** [comment]
>> Would it be ok for you guys if I ship what we have right now as 1.1.0

Absolutely, this has been my scope all the way - just document the limitation in the readme.

**Comment by jmezach** [comment]
This has been released in [1.1.0](https://www.nuget.org/packages/MSBuild.Sdk.SqlProj/1.1.0). README has been updated.
